src/algebra/order/to_interval_mod.lean

```diff
@@ -37,6 +37,41 @@ Unify `smodeq` and `add_comm_group.modeq`, which were originally developed indep
 
 noncomputable theory
 
+namespace add_comm_group
+variables {α : Type*} [add_comm_group α]
+
+/-- `a ≡ b [PMOD p]` means that `b` is congruent to `a` modulo `p`.
+
+Equivalently (as shown further down the file), `b` does not lie in the open interval `(a, a + p)`
+modulo `p`, or `to_Ico_mod hp a` disagrees with `to_Ioc_mod hp a` at `b`, or `to_Ico_div hp a`
+disagrees with `to_Ioc_div hp a` at `b`. -/
+def modeq (p a b : α) : Prop := ∃ z : ℤ, b = a + z • p
```

Comment:
This definition is probably easiest to use, but would be a departure from `nat.modeq`, `int.modeq`, and `smodeq` which are defined to be equality after the mod (%) operation or equality in the quotient. Is this what @YaelDillies plan to use as the general definition?

Reply:
People have complained in the past that `int.modeq` was defined using `%` and that `nat.modeq` wasn't defined in terms of `int.modeq`. My plan of action is to
1. define `add_comm_group.modeq`. Easy
2. drop `int.modeq` in favour of it. Probably easy, just have to realign some names and fix a few proofs (most proofs don't use the definition of `int.modeq` because it is terrible)
3. redefine `nat.modeq` in terms of it. Easy.
4. replace `add_comm_group.modeq` with notation around `smodeq`. Potentially hard because `add_comm_group.modeq` is the special case of `smodeq` where the submodule is generated by a single element.

It's likely we can only do step 1 before the port. Also note that if we go through all those steps, the definition of `int.modeq` will change twice, and hopefully that's okay because as said above almost no proof relies on the defeq because of how useless it is.

Note however that my new definition of `add_comm_group.modeq` isn't quite the same as the one Eric has here. I instead have
```
def modeq (p a b : α) : Prop := ∃ z : ℤ, b - a = z • p
```
which is defeq to `p | b - a` when `p a b : ℤ`. You can see my progress on the branch [`eric-wieser/redef-Imodeq`](https://github.com/leanprover-community/mathlib/tree/eric-wieser/redef-Imodeq).

Reply:
This sounds like a good plan, and your `add_comm_group.modeq` is also defeq to the [proposed new `smodeq` definition](https://github.com/leanprover-community/mathlib/pull/15915/files#diff-2247887f999a2fd39e7c44eb89f7db7c3b9636b7d41edfac9bd962aea97cb001R24-R25) applied to `zmultiples`, so maybe step 4 won't be hard once we adopt the new `smodeq`.

For future reference, let me note that `smodeq` (and `add_comm_group.modeq`) could be generalized further from `add_comm_group` to `add_group` to mean that two elements lie in the same [right_coset](https://leanprover-community.github.io/mathlib_docs/group_theory/coset.html#right_coset) (to express that two elements lie in the same left coset, we may say their inverses lie in the same right coset), and could be `to_additive`'d from the multiplicative version.

Reply:
Oh indeed! Then if @eric-wieser agrees, I suggest we merge `eric-wieser/redef-Imodeq` into this PR.

Reply:
It sounds like you both have stronger opinions on this than me. I propose we either:
1. Merge this PR as is, and let @YaelDillies create a follow up from the branch above
2. Abandon this PR and let @YaelDillies do the whole thing directly from the branch above
3. Abandon this PR and refrain from doing anything until mathlib4

Reply:
I'm happy to go with 2 if you're tired of leading the charge.

Reply:
#18958

Reply:
@eric-wieser @YaelDillies Does that mean that this PR should now be closed?

Reply:
Yes, let's close it.